File: scripts/apply_sql.ts
/**
 * Applique les fichiers SQL de db/ dans l'ordre.
 *
 * Equivalent Node de scripts/apply_sql.py, sans dependance Python : le projet
 * tourne deja sous Node, exiger une seconde chaine d'outils pour les
 * migrations n'apporte rien.
 *
 *     node --env-file=.env scripts/apply_sql.ts --dry-run
 *     node --env-file=.env scripts/apply_sql.ts
 *     node --env-file=.env scripts/apply_sql.ts --only 004_client_space.sql
 *     node --env-file=.env scripts/apply_sql.ts --file db/dev/reset.sql
 */
import { readFileSync, readdirSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

import pg from 'pg';

const ROOT = join(dirname(fileURLToPath(import.meta.url)), '..');
const args = process.argv.slice(2);
const flag = (name: string) => {
  const i = args.indexOf(name);
  return i >= 0 ? (args[i + 1] ?? '') : undefined;
};

const only = flag('--only');
const single = flag('--file');
const dryRun = args.includes('--dry-run');

let files: string[];
if (single) {
  files = [resolve(ROOT, single)];
} else {
  // Volontairement non recursif : db/dev/ et db/examples/ ne doivent jamais
  // partir par accident lors d'une application normale.
  files = readdirSync(join(ROOT, 'db'))
    .filter((f) => f.endsWith('.sql'))
    .sort()
    .map((f) => join(ROOT, 'db', f));
  if (only) files = files.filter((f) => f.endsWith(only));
}

if (files.length === 0) {
  console.error('aucun fichier a appliquer');
  process.exit(1);
}

if (dryRun) {
  console.log('fichiers qui seraient appliques, dans cet ordre :');
  for (const f of files) {
    const lines = readFileSync(f, 'utf8').split('\n').length;
    console.log(`  ${f.split(/[\\/]/).pop()!.padEnd(28)} ${String(lines).padStart(4)} lignes`);
  }
  process.exit(0);
}

const dsn = process.env.DATABASE_URL;
if (!dsn) {
  console.error('DATABASE_URL non defini');
  process.exit(1);
}

const client = new pg.Client({ connectionString: dsn, ssl: { rejectUnauthorized: false } });
await client.connect();

const info = await client.query('SELECT current_database() d, current_user u, version() v');
console.log(`connecte  base=${info.rows[0].d}  role=${info.rows[0].u}`);
console.log(`          ${String(info.rows[0].v).split(',')[0]}\n`);

let failed = false;
for (const f of files) {
  const name = f.split(/[\\/]/).pop()!;
  try {
    // Chaque fichier porte son propre BEGIN/COMMIT : il s'applique en entier
    // ou pas du tout.
    await client.query(readFileSync(f, 'utf8'));
    console.log(`  OK      ${name}`);
  } catch (err) {
    failed = true;
    console.log(`  ECHEC   ${name}`);
    console.log(`          ${err instanceof Error ? err.message : String(err)}`);
    break;
  }
}

await client.end();
console.log(failed ? '\nInterrompu. Corriger puis relancer.' : '\nSchema applique.');
process.exit(failed ? 1 : 0);

File: scripts/apply_sql.py
#!/usr/bin/env python3
"""
Applique les fichiers SQL de db/ dans l'ordre, sans avoir psql installe.

    pip install "psycopg[binary]"
    export DATABASE_URL="postgresql://postgres.<ref>:<motdepasse>@aws-1-eu-west-1.pooler.supabase.com:5432/postgres"
    python scripts/apply_sql.py --dry-run
    python scripts/apply_sql.py

Sur Supabase, prendre la chaine "Session pooler" (port 5432) dans
Project Settings -> Database, pas la "Transaction pooler" (6543) : le mode
transaction ne supporte ni les instructions preparees ni tout ce qui tient sur
la duree d'une session, ce qui gene les migrations.
Les instructions preparees sont desactivees ci-dessous par precaution, pour que
le script fonctionne aussi via le pooler en mode transaction.
"""

from __future__ import annotations

import argparse
import glob
import os
import sys

HERE = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--dry-run", action="store_true",
                    help="liste les fichiers sans se connecter")
    ap.add_argument("--only", help="n'appliquer qu'un fichier de db/, ex. 003_rls_supabase.sql")
    ap.add_argument("--file", help="appliquer un chemin precis, y compris hors de db/ "
                                   "(ex. db/dev/reset.sql, db/examples/seed_fitzone.sql)")
    args = ap.parse_args()

    if args.file:
        path = args.file if os.path.isabs(args.file) else os.path.join(HERE, args.file)
        if not os.path.isfile(path):
            sys.exit(f"fichier introuvable : {args.file}")
        files = [path]
    else:
        # Glob volontairement non recursif : db/dev/ et db/examples/ ne doivent
        # jamais partir par accident.
        files = sorted(glob.glob(os.path.join(HERE, "db", "*.sql")))
        if args.only:
            files = [f for f in files if os.path.basename(f) == args.only]
            if not files:
                sys.exit(f"fichier introuvable : {args.only}")

    if args.dry_run:
        print("fichiers qui seraient appliques, dans cet ordre :")
        for f in files:
            n = sum(1 for line in open(f, encoding="utf-8"))
            print(f"  {os.path.basename(f):<28} {n:>4} lignes")
        return 0

    dsn = os.environ.get("DATABASE_URL")
    if not dsn:
        sys.exit("DATABASE_URL non defini")

    try:
        import psycopg
    except ImportError:
        sys.exit('psycopg manquant :  pip install "psycopg[binary]"')

    # autocommit : chaque fichier porte son propre BEGIN/COMMIT, on le laisse
    # piloter ses transactions plutot que d'en ouvrir une par-dessus.
    with psycopg.connect(dsn, autocommit=True, prepare_threshold=None) as conn:
        cur = conn.cursor()
        cur.execute("SELECT current_database(), current_user, version()")
        db, user, version = cur.fetchone()
        print(f"connecte  base={db}  role={user}")
        print(f"          {version.split(',')[0]}")
        print()

        for f in files:
            name = os.path.basename(f)
            sql = open(f, encoding="utf-8").read()
            try:
                cur.execute(sql)
                print(f"  OK      {name}")
            except Exception as exc:
                print(f"  ECHEC   {name}")
                print(f"          {type(exc).__name__}: {exc}")
                print("\nRien de ce fichier n'a ete applique (son BEGIN/COMMIT "
                      "garantit le tout-ou-rien). Corriger puis relancer.")
                return 1

    print("\nschema applique.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
